### ISSUE TITLE
Database engine testing by the netdata team

### ISSUE THREAD

**Issue by mfundul** [netdata/netdata#5878]
Performance and functional testing of the new DB engine (#5282) by the netdata team:
- [x] Functional testing (everything works as it should)
- [x] Black box testing using netdata APIs (e.g. /api/v1/data, backends etc.) @cakrit
- [x] Building/packaging
- [x] Run tests on multiple infrastructures

___

Instructions: [link](https://github.com/netdata/netdata/pull/5282#issuecomment-483998764)

 - install dependencies
   - `libuv` multi-platform support library with a focus on asynchronous I/O
   - `liblz4` extremely Fast Compression algorithm
   - `Judy` general purpose dynamic array
   - `openssl` cryptography and SSL/TLS Toolkit
 - install dbengine branch `https://github.com/mfundul/netdata --branch=database-engine`
 - enable the new DB engine in the configuration file

```bash
[global]
  memory mode = dbengine
```
 - restart `netdata.service`, profit!

Run workload generator:
```
/usr/sbin/netdata  -W createdataset=3600
```
Where 3600 is seconds, you can set it to any time.

**Comment by thiagoftsm** [comment]
On Slackware Linux Current, everything looks line fine since the compilation until now after few hours running.

On FreeBSD on the other hand I have to install the following lib to put the Netdata running:

Uuid: The Netdata script is calling this lib "uuid-dev" (or "libuuid-devel"), but on FreeBSD it is known in the ports as misc/e2fsprogs-libuuid.

There is another important point that I would like to call attention with relationship to the libraries. I ran ldd on netdata compiled on Slackware and on BSD:

Slackware$ ldd /usr/sbin/netdata
        linux-vdso.so.1 ()
        libcap.so.2 => /lib64/libcap.so.2 ()
        libm.so.6 => /lib64/libm.so.6 ()
        libz.so.1 => /lib64/libz.so.1 ()
        libuuid.so.1 => /lib64/libuuid.so.1 ()
        libuv.so.1 => /usr/lib64/libuv.so.1 ()
        libJudy.so.1 => /usr/lib64/libJudy.so.1 ()
        libcrypto.so.1.1 => /lib64/libcrypto.so.1.1 ()
        libssl.so.1.1 => /lib64/libssl.so.1.1 ()
        libpthread.so.0 => /lib64/libpthread.so.0 ()
        libc.so.6 => /lib64/libc.so.6 ()
        /lib64/ld-linux-x86-64.so.2 ()
        librt.so.1 => /lib64/librt.so.1 ()
        libnsl.so.1 => /lib64/libnsl.so.1 ()
        libdl.so.2 => /lib64/libdl.so.2 ()

FreeBSD $ ldd /usr/sbin/netdata
        libm.so.5 => /lib/libm.so.5 ()
        libz.so.6 => /lib/libz.so.6 ()
        libuuid.so.1 => /usr/local/lib/libuuid.so.1 ()
        libcrypto.so.111 => /lib/libcrypto.so.111 ()
        libssl.so.111 => /usr/lib/libssl.so.111 ()
        libthr.so.3 => /lib/libthr.so.3 ()
        libc.so.7 => /lib/libc.so.7 (0x8008e6000)

And after I install the libuv and libJudy, the script did not recognize it and the compilation was done without both libraries.

**Comment by mfundul** [comment]
@thiagoftsm please provide the output of the `./configure` part of the installation so that we may understand what went wrong. It might be that libuv is version 0 whereas we need libuv version 1.

**Comment by mfundul** [comment]
So on CentOS 6 libuv does not seem to work:
```
checking for uv_fs_scandir_next in -luv... no
```
libuv version is 0.10 instead of v1

**Comment by mfundul** [comment]
> 
> 
> So on CentOS 6 libuv does not seem to work:
> 
> ```
> checking for uv_fs_scandir_next in -luv... no
> ```
> 
> libuv version is 0.10 instead of v1

It's in OKay [repository](https://okay.network/blog-news/rpm-repositories-for-centos-6-and-7.html).

**Comment by mfundul** [comment]
We require libuv version >= 1.16

**Comment by thiagoftsm** [comment]
On FreeBSD 12.0-RELEASE-p3 it ran without any problem during few hours. I only had to install the following packages: liblz4 e2fsprogs-libuuid libuv go Judy

On Slackware it was necessary to install the liblz4, the other libs were already present, but when I tried to run the netdata with the necessary configuration to use the new engine, it always died with messages like this:

2019-04-18 12:45:26: go.d CRITICAL: main[main] SIGPIPE received. Terminating...

Firstly it is important to say that when we do not use the new engine it works fine on Slackware current, but our goal is to have the new engine running, so I decided to compare the version of each go language compiler and I also tested on CentOS. I am copying below the outputs

centos$ go version
go version go1.11.5 linux/amd64

freebsd$ go version
go version go1.12.1 freebsd/amd64

slackware$go version
go version go1.10.3 gccgo (GCC) 8.3.0 linux/amd64

When I saw these output, I decided to compile the go language 1.11.4 on Slackware and after this everything ran fine, so it looks like that we would need at least the go 1.11.4 to run the new engine.

**Comment by vlvkobal** [comment]
There are a lot of warning messages when netdata is being built with
```
sudo CFLAGS="-O1 -ggdb -Wall -Wextra -Wformat-signedness -fstack-protector-all -DNETDATA_INTERNAL_CHECKS=1 -D_FORTIFY_SOURCE=2 -DNETDATA_VERIFY_LOCKS=1" ./netdata-installer.sh --enable-plugin-nfacct --enable-plugin-freeipmi --disable-go --disable-lto --dont-wait
```

**Comment by vlvkobal** [comment]
Static binary build is not working:
```
2019-04-19 14:05:51: netdata FATAL : MAIN :RRD_MEMORY_MODE_DBENGINE is not supported in this platform. # : Invalid argument

2019-04-19 14:05:51: netdata INFO  : MAIN : /opt/netdata/usr/libexec/netdata/plugins.d/anonymous-statistics.sh 'FATAL' 'netdata:MAIN' '0246@database/r:rrdhost_create /22'
2019-04-19 14:05:51: netdata INFO  : MAIN : EXIT: netdata prepares to exit with code 1...
2019-04-19 14:05:51: netdata INFO  : MAIN : /opt/netdata/usr/libexec/netdata/plugins.d/anonymous-statistics.sh 'EXIT' 'ERROR' '-'
2019-04-19 14:05:52: netdata INFO  : MAIN : EXIT: cleaning up the database...
2019-04-19 14:05:52: netdata INFO  : MAIN : Cleaning up database [0 hosts(s)]...
```

**Comment by thiagoftsm** [comment]
On 32-bits computer, when we use clang 6.0.1 compiler(FreeBSD 12.0-RELEASE-3) it stops the compilation with the errors like this:

global_statistics.c:(.text+0x32): undefined reference to `__atomic_fetch_add_8'
global_statistics.c:(.text+0x4e): undefined reference to `__atomic_fetch_add_8'
global_statistics.c:(.text+0x6d): undefined reference to `__atomic_fetch_add_8'

But, on the same 32 bits computer when I used GCC 8.3.0(Slackware Current), I did not have any problem in the compilation and the new engine works fine.

**Comment by ilyam8** [comment]
Doesn't compile on debian9

```
database/engine/datafile.c: In function 'create_data_file':
database/engine/datafile.c:103:39: error: 'UV_FS_O_DIRECT' undeclared (first use in this function)
     fd = uv_fs_open(NULL, &req, path, UV_FS_O_DIRECT | UV_FS_O_CREAT | UV_FS_O_RDWR | UV_FS_O_TRUNC,
                                       ^~~~~~~~~~~~~~
database/engine/datafile.c:103:39: note: each undeclared identifier is reported only once for each function it appears in
database/engine/datafile.c:103:56: error: 'UV_FS_O_CREAT' undeclared (first use in this function)
     fd = uv_fs_open(NULL, &req, path, UV_FS_O_DIRECT | UV_FS_O_CREAT | UV_FS_O_RDWR | UV_FS_O_TRUNC,
                                                        ^~~~~~~~~~~~~
database/engine/datafile.c:103:72: error: 'UV_FS_O_RDWR' undeclared (first use in this function)
     fd = uv_fs_open(NULL, &req, path, UV_FS_O_DIRECT | UV_FS_O_CREAT | UV_FS_O_RDWR | UV_FS_O_TRUNC,
                                                                        ^~~~~~~~~~~~
database/engine/datafile.c:103:87: error: 'UV_FS_O_TRUNC' undeclared (first use in this function)
     fd = uv_fs_open(NULL, &req, path, UV_FS_O_DIRECT | UV_FS_O_CREAT | UV_FS_O_RDWR | UV_FS_O_TRUNC,
                                                                                       ^~~~~~~~~~~~~
database/engine/datafile.c: In function 'load_data_file':
database/engine/datafile.c:186:39: error: 'UV_FS_O_DIRECT' undeclared (first use in this function)
     fd = uv_fs_open(NULL, &req, path, UV_FS_O_DIRECT | UV_FS_O_RDWR, S_IRUSR | S_IWUSR, NULL);
                                       ^~~~~~~~~~~~~~
database/engine/datafile.c:186:56: error: 'UV_FS_O_RDWR' undeclared (first use in this function)
     fd = uv_fs_open(NULL, &req, path, UV_FS_O_DIRECT | UV_FS_O_RDWR, S_IRUSR | S_IWUSR, NULL);
                                                        ^~~~~~~~~~~~
Makefile:2606: recipe for target 'database/engine/datafile.o' failed
make[2]: *** [database/engine/datafile.o] Error 1
make[2]: Leaving directory '/home/test/test/netdata'
Makefile:2712: recipe for target 'all-recursive' failed
make[1]: *** [all-recursive] Error 1
make[1]: Leaving directory '/home/test/test/netdata'
Makefile:1444: recipe for target 'all' failed
make: *** [all] Error 2

```

> Package: libuv1
Version: 1.18.0-3~bpo9+1
Installed-Size: 227


> Package: libjudydebian1
Source: judy
Version: 1.0.5-5


> Package: liblz4-1
Source: lz4 (0.0 r131-2)
Version: 0.0 r131-2+b1

> Package: openssl
Version: 1.1.0j-1~deb9u1
Installed-Size: 1309

**Comment by mfundul** [comment]
We currently require libuv version >= 1.16, I probably have to do some porting cause this is the same problem as with CentOS 6.

**Comment by ilyam8** [comment]
> We currently require libuv version >= 1.16

i hope it is `currently` otherwise it won't work on debian7,8,9, centos6, ubuntu16 (deb9 is top2 OS according to anon stats, deb8 in top10) by default.

**Comment by mfundul** [comment]
> 
> 
> > We currently require libuv version >= 1.16
> 
> i hope it is `currently` otherwise it won't work on debian7,8,9, centos6, ubuntu16 (deb9 is top2 OS according to anon stats, ubuntu16 top4, deb8 in top10) by default.

I just pushed a fix, can you check if it works for debian9?

**Comment by ilyam8** [comment]
yes

**Comment by thiagoftsm** [comment]
@mfundul I was verifying the official documentation to check, and I saw that UV_FS_O_DIRECTis not compatible with MAC OS too ( http://docs.libuv.org/en/v1.x/fs.html ).

**Comment by cakrit** [comment]
First installation with a master/slave set up. Stopped netdata, removed all contents of `/var/cache/netdata` and started it again.
- Directories for each master node chart are created under  `/var/cache/netdata`. They shouldn't be, they are empty. Expected behavior: Just see a `dbengine` folder under `/var/cache/netdata`
- Directory created as expected for the slave node, but it somehow uses the legacy DB. No `dbengine` folder under the created directory `/var/cache/netdata/{slave machine GUID}`. How/where is it that we specify the dbengine for each slave?

**Comment by mfundul** [comment]
> 
> 
> First installation with a master/slave set up. Stopped netdata, removed all contents of `/var/cache/netdata` and started it again.
> 
>     * Directories for each master node chart are created under  `/var/cache/netdata`. They shouldn't be, they are empty. Expected behavior: Just see a `dbengine` folder under `/var/cache/netdata`
> 
>     * Directory created as expected for the slave node, but it somehow uses the legacy DB. No `dbengine` folder under the created directory `/var/cache/netdata/{slave machine GUID}`. How/where is it that we specify the dbengine for each slave?

The expected behaviour is not just the `dbengine` folder. Netdata uses chart metadata that are not part of the DB engine and they stay there until we implement something like the metrics map in netdata.

As far as how we configure `memory mode` for RRD hosts I don't know, I've never done this before. Ideally it should be already possible.

**Comment by cakrit** [comment]
The setting `page cache size = 32` means that we allocate 32GB of memory? 

If yes, which one is the page size setting and why do I see the high memory usage below?
![Screenshot from 2019-05-01 20-38-39](https://user-images.githubusercontent.com/43294513/57035213-c2ed6780-6c51-11e9-9376-6e2caf923d93.png)

If not, how do I control how much memory I use? Is it the old `history` setting? We really need the documentation of the DB config options.

**Comment by mfundul** [comment]
> 
> 
> The setting `page cache size = 32` means that we allocate 32GB of memory?
> 
> If yes, which one is the page size setting and why do I see the high memory usage below?
> ![Screenshot from 2019-05-01 20-38-39](https://user-images.githubusercontent.com/43294513/57035213-c2ed6780-6c51-11e9-9376-6e2caf923d93.png)
> 
> If not, how do I control how much memory I use? Is it the old `history` setting? We really need the documentation of the DB config options.

Page cache size is in MiB. The allocated memory I see is 70MiB. It does not look like high memory usage.

**Comment by cakrit** [comment]
> The expected behaviour is not just the `dbengine` folder. Netdata uses chart metadata that are not part of the DB engine and they stay there until we implement something like the metrics map in netdata.

But those directories are empty. They only hold metrics, when we do not use the new dbengine. Or is it that we use the directory names themselves for the metadata you mentioned? Because that's just the chart name. 

> As for how we configure `memory mode` for RRD hosts I don't know, I've never done this before. Ideally it should be already possible.

We didn't have a way to specify different memory modes per slave (host). This is a new feature. If you haven't specified a way to override the default (master) config for slaves, then they should all use the same settings as the master (e.g. same page size).

**Comment by cakrit** [comment]
> We didn't have a way to specify different memory modes per slave (host). I am talking about how the master saves the replicated data of course, not what memory mode the slaves use themselves. This is a new feature. If you haven't specified a way to override the default (master) config for slaves, then they should all use the same settings as the master (e.g. same page size and dbengine). The proper place to put these settings is in each slave's `stream.conf`, but that requires that the streaming protocol is updated to accept the DB info as well.

Actually, I hadn't noticed, but we do have this feature already. The master node's `stream.conf` can be configured as follows:

```
    # The default memory mode to be used for all hosts using this API key.
    # You can also set it per host below.
    # If you don't set it here, the memory mode of netdata.conf will be used.
    # Valid modes:
    #    save    save on exit, load on start
    #    map     like swap (continuously syncing to disks - you need SSD)
    #    ram     keep it in RAM, don't touch the disk
    #    none    no database at all (use this on headless proxies)
    default memory mode = dbengine
```

**Comment by mfundul** [comment]
I just pushed a fix for multiple slaves pushing to a master that has many DB engine instances crashing with segmentation fault.

About the empty directories, I just looked at the code and noticed that the behaviour is identical to RRD_MEMORY_MODE_RAM. If I should fix this for RRD_MEMORY_MODE_DBENGINE then I must do the exact same thing for RRD_MEMORY_MODE_RAM as well. Moreover, this may be a problem for users that switch memory modes and I tend to not want to mess with this situation.

**Comment by vlvkobal** [comment]
To install `libJudy 1.0.5` on macOS:
```
brew pull 13404
brew install judy
```
After installing openssl:
```
ln /usr/local/Cellar/openssl/1.0.2r/lib/libcrypto.a /usr/local/lib/
ln -s /usr/local/Cellar/openssl/1.0.2r/include/openssl/ /usr/local/include
```

I've sent the installation log with compilation errors to @mfundul

**Comment by thiagoftsm** [comment]
I remember this afternoon what I had to change in the install on FreeBSD. The new libraries were installed inside the directory /usr/local/lib and the compiler could not find them, so I changed the line 385 to:

CFLAGS="${CFLAGS}" LDFLAGS="-L/usr/local/lib" || exit 1

**Comment by vlvkobal** [comment]
@mfundul already committed this workaround for FreeBSD: https://github.com/netdata/netdata/pull/5282/commits/b181136c1cfa41fec56521b69f7be10b83143fdd

**Comment by mfundul** [comment]
Since `O_DIRECT` (direct I/O in `open()`) does not exist in OS X I'm posting [this](https://github.com/axboe/fio/issues/48) for reference.
